### PR TITLE
Gemma2 and flash-attention

### LINKED ISSUE #31953: Flash Attention with Gemma 2
### System Info

- `transformers` version: 4.43.0.dev0
- Platform: Windows-10-10.0.22631-SP0
- Python version: 3.11.9
- Huggingface_hub version: 0.23.4
- Safetensors version: 0.4.3
- Accelerate version: 0.32.1
- Accelerate config:
        - compute_environment: LOCAL_MACHINE
        - distributed_type: NO
        - mixed_precision: bf16
        - use_cpu: False
        - debug: False
        - num_processes: 1
        - machine_rank: 0
        - num_machines: 1
        - gpu_ids: all
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - enable_cpu_affinity: False
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []
- PyTorch version (GPU?): 2.3.0+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: No
- Using GPU in script?: Yes
- GPU type: NVIDIA GeForce RTX 4070
- Flash attention = 2.6.1 (Added by me)

### Who can help?

@ArthurZucker

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

```python
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig
import torch
import os
import warnings
import timeit

def start_timing(process="process"):
    global start_time
    start_time = timeit.default_timer()
    print(f"Started {process}.")

def finish_timing(description, suffix=""):
    global start_time
    end_time = timeit.default_timer()
    elapsed_time = end_time - start_time
    print(f"{description} took {elapsed_time:.2f}s {suffix}.")

warnings.filterwarnings("ignore")

os.environ["USE_FLASH_ATTENTION"]="1"

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16
)

checkpoint = "google/gemma-2-9b"

tokenizer = AutoTokenizer.from_pretrained(checkpoint)

model = AutoModelForCausalLM.from_pretrained(checkpoint, device_map="cuda", quantization_config=bnb_config, torch_dtype='bfloat16')

input_text = \
"""Hi, my name is"""

base_output = "idk"

context = ""

for line in input_text.split("\n"):
    context += line+" "
    for optim in ["eager", "sdpa", "flash_attention_2"]: #
        torch.manual_seed(0)
        torch.cuda.empty_cache()
        input_ids = tokenizer(context, return_tensors="pt", add_special_tokens=True).to("cuda")
        model.config._attn_implementation = optim
        start_timing()
        outputs = model.generate(**input_ids, max_new_tokens=100, pad_token_id=tokenizer.eos_token_id)
        finish_timing(f"The {optim} attention implementation", f"to infer 100 tokens")
        if optim == "eager":
            base_output = tokenizer.decode(outputs[0])
        if base_output == tokenizer.decode(outputs[0]):
            print(f"The {optim} attention returned the same output as eager attention.")
        else:
            print(f"The {optim} attention didn't return the same output as eager attention.")
        print()
```

### Expected behavior

The flash attention implementation isn't returning same output as eager.

Sometimes, the results are same, but sometimes, (especially in long context), it fails totally.

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Fixes #31953. As explained in the comments section, the issue with flash attn was static cache, as we were not preparing special causal masks with padding on both sides, for pad-token and for cache placeholders.

I tried to pass-in a 2D mask with pads for cache positions, yet it didn't work out because when unpadding for flash attention we fall into the path when query length is not same as key length, and remain with only the leftmost 1 query token in first forward.

So I think we can use a similar approach to the one discussed in one prev PR to cut out memory usage, cut out key/values to remove extra padding on the right

Plus, there were cache related issue on Gemma2, I will make another PR on that

### CODE REVIEW

File: src/transformers/models/gemma2/modeling_gemma2.py [hunk 1]
```diff
@@ -324,6 +324,11 @@ def forward(
             }
             key_states, value_states = past_key_value.update(key_states, value_states, self.layer_idx, cache_kwargs)
 
+        if attention_mask is not None:
+            seq_len = attention_mask.shape[1]
+            key_states = key_states[:, :, :seq_len]
+            value_states = value_states[:, :, :seq_len]
```
Comment (ArthurZucker) [inline src/transformers/models/gemma2/modeling_gemma2.py:330]: ah, yeah static cache does not support that because it's input dependent, but great catch

**Review by ArthurZucker — APPROVED** [review]
Good catch indeed. Since we don't have attention mask we can't say let's not take care the padding into account!

**Review by gante — APPROVED** [review]
Thank you for fixing 🙌

**Comment by Oxi84** [comment]
It works but as far as I can see the vram usage is the same, and the speed is a bit slower than without attention. There might still be something broken, in sense that it does not actually use the flash_attention_2 at all.


    base_model = AutoModelForCausalLM.from_pretrained(
        base_model_id, 
        quantization_config=bnb_config, 
        device_map={"": 0},
        trust_remote_code=True,
        attn_implementation="flash_attention_2",
    )

     with torch.no_grad():
                outputs = ft_model.generate(**model_inputs, max_new_tokens=60, do_sample=True, top_k=100, top_p=1.0,
                        pad_token_id=tokenizer.pad_token_id,
                        use_cache=True,
                        )

**Comment by zucchini-nlp** [comment]
@Oxi84 hmm, indeed Gemma2 doesn't benefit much from Flash-attention compared to other models. Seems like it's related to teh big vocab size it has (256000 tokens) which we have to store all in fp32 dtype for sampling. In pre-fill stage that is a tensor of shape (bs, seq-len, 256000) and needs a huge memory for long sequences. 

We can hugely benefit from https://github.com/huggingface/transformers/pull/31292  in Gemma2 after it's merged

**Comment by zucchini-nlp** [comment]
Will be merging today at night

File: tests/models/gemma2/test_modeling_gemma2.py [hunk 2]
```diff
@@ -161,3 +164,29 @@ def test_model_9b_pipeline_bf16(self):
...
         self.assertEqual(output[1][0]["generated_text"], EXPECTED_TEXTS[1])
+
+    @require_read_token
+    @require_flash_attn
+    @require_torch_gpu
+    @mark.flash_attn_test
+    @slow
+    def test_model_9b_flash_attn(self):
+        # See https://github.com/huggingface/transformers/issues/31953 --- flash attn was generating garbage for gemma2, especially in long context
+        # NOTE: the quality is a lot better whan fp16 is used, and worse for bf16
```
Comment (tanliboy) [inline tests/models/gemma2/test_modeling_gemma2.py:175]: Could you elaborate more on this? Specifically, does using `flash_attention_2` negatively impact the quality of the model's output with `bfloat16` compared to using `eager` mode?

Comment (zucchini-nlp) [reply tests/models/gemma2/test_modeling_gemma2.py]: Oh, this was true until I discovered a bug I made. It doesn't matter much which precision is used, removed the comment
